Add param completion command

rojepp opened the pull request:
This is a starting point for #29.

I'm not sure I'm happy with the serialized format, see added `output.json`.

rojepp committed:
Methods WIP
Some intergration tests currently fail, will investigate.

rojepp committed:
Fix integration tests

rojepp committed:
Fix test diffs on my machine by sorting references on filename only

rojepp committed:
Add missing output.json

rojepp commented on FSharp.AutoComplete/Program.fs:
```diff
@@ -366,7 +370,7 @@ module internal Main =
     |]
 
-  let prAsJson o = printAgent.WriteLine (JsonConvert.SerializeObject(o, jsonConverters))
+  let prAsJson o = printAgent.WriteLine (JsonConvert.SerializeObject(o, Formatting.Indented,  jsonConverters))
```
This prints indented, instead of trying to indent later in integration tests. Hope this is ok?

rneatherway replied:
I'm afraid not really. The JSON parsers that interpret what is printed by FSAC assume that each message is exactly one line.

rojepp replied:
I've reverted this.

rojepp commented on FSharp.AutoComplete/Program.fs:
```diff
@@ -472,7 +476,7 @@ module internal Main =
                                           Files = files
                                           Output = targetFilename
-                                          References = List.sort p.References
+                                          References = List.sortBy Path.GetFileName p.References
```
Sort by file name only, fixes the issue with references appearing out of order on my machine, sensitive to user names, possibly machine name, and other paths(?)

rneatherway replied:
Sounds good

rojepp commented on FSharp.AutoComplete/test/integration/TestHelpers.fsx:
```diff
@@ -81,9 +84,10 @@ let writeNormalizedOutput (fn: string) (s: string) =
+    
+    // This throws because each line is not valid json by itself
+    //if Path.GetExtension fn = ".json" then
+    //  lines.[i] <- formatJson lines.[i]
```
Had to disable this, it was throwing exceptions.

rneatherway replied:
Each line _should_ be valid json. Did this happen before you changed `Formatting.Indented` above?

rojepp committed:
Revert json indent settings

rneatherway commented on FSharp.AutoComplete/Program.fs:
```diff
@@ -607,6 +611,21 @@ module internal Main =
 
+
+          | Methods ->
+            let meth = tyRes.GetMethods(line, col, lineStr)
```
I think at this point we should first scan back using [this code](https://github.com/fsharp/xamarin-monodevelop-fsharp-addin/blob/master/MonoDevelop.FSharpBinding/FSharpTextEditorCompletion.fs#L318). An editor should be able to give the current cursor position with some parameters already filled in.

rneatherway replied:
I wonder on this point, though, with multiple parameters how we could communicate to the editor which parameter we are on.

rojepp replied:
Maybe something to consider when deciding on a json format. Which parameter of which overload.

rneatherway commented on FSharp.AutoComplete/Program.fs:
```diff
@@ -607,6 +611,21 @@ module internal Main =
+            let meth = tyRes.GetMethods(line, col, lineStr)
+                       |> Async.RunSynchronously
+            match meth with
+            | Some (name,info) ->
```
info.Length should be greater than 0 (https://github.com/fsharp/xamarin-monodevelop-fsharp-addin/blob/master/MonoDevelop.FSharpBinding/FSharpTextEditorCompletion.fs#L351)

rojepp replied:
Done

rneatherway commented:
I think the test should have some overloads and multiple parameters. Perhaps something like:

```
type NewObjectType() =

  member x.Terrific (y : int) (z : char) : int =  y
  member x.Terrific (y : int) (z : System.DateTime) : int =  y
  member x.Terrific (y : Set<'a>) (z : int) : Set<'a> =  y
```

And maybe using something from the standard library like `System.DateTime.Parse`. I could send a PR to this PR if that'll help?

rneatherway commented:
I agree, I don't think the structure with tuples and ADTs has serialized well at all. We should see how it looks in a more complicated case and come up with a record and list structure that can be consumed cleanly.

rojepp committed:
Guard for info.Length = 0

rojepp commented:
Please do send a PR if you'd like to help out!

rojepp commented:
I'd appreciate a test that fails because the startOffset code is not there.

rneatherway committed:
Test with overloaded methods and request in call

rojepp committed:
Merge pull request #1 from rneatherway/methods

Test with overloaded methods and request in call

rojepp committed:
Incorporate finding of start position from fsharbinding

rojepp commented:
@rneatherway I've incorporated the startPos code from fsharpbinding, and the tests pass. 
What are your thoughts on the output format?

rneatherway commented:
Thanks, I will have a think about it tomorrow. I definitely think we should define a simple type hierarchy to return the results in.

rojepp commented:
Given

``` json
{
  "Kind": "method",
  "Data": {
    "Item1": "NewObjectType",
    "Item2": [
      {
        "Description": {
          "Case": "FSharpToolTipText",
          "Fields": [
            [
              {
                "Case": "Group",
                "Fields": [
                  [
                    {
                      "Item1": "new : unit -> FileTwo.NewObjectType",
                      "Item2": {
                        "Case": "None"
                      }
                    }
                  ]
                ]
              }
            ]
          ]
        },
        "TypeText": ": FileTwo.NewObjectType",
        "Parameters": [],
        "IsStaticArguments": false
      }
    ]
  }
}
```

``` fsharp
type OverloadParameter =
  {
    ParameterName : string
    Name : string
    CanonicalTypeTextForSorting : string
    Display : string
    Description : string
  }
type Overload = 
  {
    Tip : string
    TypeText : string
    Parameters : OverloadParameter list
    IsStaticArguments : bool
  }
type MethodResponse =
  {
    Name : string
    Overloads : Overload list
  }
```

There are some assumptions here: Description always returns `FSharpToolTipText` DU case, `Fields` always returns DU case `Group`. I have no idea if these assumptions hold. :) This is as flat as I could get it without losing any information. I don't presume to know exact uses of this, so I'd rather not leave anything out.

rojepp committed:
Improved json format for methods.
There may be more to do, but now it looks pretty usable.

rojepp commented:
@rneatherway There's a deliberate serialization in place. It may need improvements, but it looks pretty ok to me. I didn't add XmlDoc resolution, I consider that a separate issue. There are now two places where that should be done.

rneatherway commented:
If your assumption is just on the tooltip text, then it seems reasonable to assume that passing it to `TipFormatter.formatTip` should account for the different cases there. It's interesting, I suppose if we do want to allow editors to produce a richer tooltip output then we will want to stop called that function and return the fully structured information somehow.

rneatherway commented on FSharp.AutoComplete/test/integration/ParamCompletion/ParamCompletionRunner.fsx:
```diff
@@ -0,0 +1,30 @@
+p.parse "Program.fs"
+Threading.Thread.Sleep(6000)
+p.methods "Program.fs" 4 36
+p.methods "Program.fs" 4 37
```
This one is failing, but I think it shouldn't. Using 1-indexed columns this is a request at `$` in `let testval = FileTwo.NewObjectType($)`, so I think we should return results here. The previous one is actually on the open bracket, which it would be nice if it succeeded, but I think slightly less important than this one.

rojepp replied:
Indeed, I'd go so far as to say the first one shouldn't succeed. :)

rneatherway commented on FSharp.AutoComplete/Program.fs:
```diff
@@ -607,6 +631,64 @@ module internal Main =
+              let rec loop depth (line, col) =
+                if (line,col) <= (1,1) then (line, col) else
+                let ch = state.Files.[file].Lines.[line - 1].[col - 1]
+                if (ch = '(' || ch = '{' || ch = '[') && depth > 0 then loop (depth - 1) (prevPos (line,col))
```
This returns the character where the `(` is actually found, which doesn't seem to match the comment. However, it does match the parser it gets passed to: https://github.com/fsharp/FsAutoComplete/blob/master/FSharp.CompilerBinding/Parser.fs#L302

The `($)` example I pointed out above preferably succeed, maybe just by searching from the column _before_ the cursor rather than the one _at_ the cursor. This would be as simple as changing the match below to pass it `col-1` instead. It might be better though to use the column after the paren and adjust the parser correspondingly.

I wonder if we could bail out at some point? If the file is very large and we are not in a method call this will go back to the start of the file. I thought of seeing a `=` when `depth` is 0, but that could be part of an expression argument of type `bool`. Perhaps a cap on the number of lines to go back? Seeing a `let`? Any thoughts?

rneatherway commented:
Looking good, but I'd like to sort out this off-by-1 issue. They can be really annoying to debug at the level of the editor. The other thing is, I wonder if we could add a field to `MethodResponse` called `Parameter: int`, which indicates what parameter we are on. This would be helpful for an editor and would require extending the scanning-back loop to track the number of commas seen at the current depth.

rojepp commented:
Great feedback, thanks. I'll try and make some time to address your concerns. We should definitely bail at some arbitrary cutoff, 3 lines or so. :)

rojepp committed:
Fix off-by-one and add cutoff to avoid searching through the whole file

rojepp committed:
Report current parameter index by counting commas at depth 0.

rojepp commented:
I've addressed all concerns except this one:

> I suppose if we do want to allow editors to produce a richer tooltip output then we will want to stop called that function and return the fully structured information somehow.

Do you have a suggestion for how this might look?

rneatherway commented:
Great, thanks! I will have time to review and merge this tomorrow. At a quick glance it all looks good to me. Regarding the richer tooltip output, I don't think that is important for this change. I was just thinking of  offering the option for this and the normal tooltips of returning structured data rather than a preformatted string.

7sharp9 commented:
If you find any bugs with the source from the F# addin please PR back upstream, there are quite a few edge cases that probably are not covered.
